File: app/src/main/java/soup/movie/ui/main/MainPresenter.java
package soup.movie.ui.main;

import javax.inject.Inject;

import soup.movie.di.ActivityScoped;
import soup.movie.settings.HomeTypeSetting;
import soup.movie.ui.BasePresenter;

import static soup.movie.ui.main.MainContract.TAB_MODE_NOW;
import static soup.movie.ui.main.MainContract.TAB_MODE_PLAN;
import static soup.movie.ui.main.MainContract.TAB_MODE_SETTINGS;

@ActivityScoped
public class MainPresenter extends BasePresenter<MainContract.View>
        implements MainContract.Presenter {

    private final HomeTypeSetting homeTypeSetting;

    @Inject
    MainPresenter(HomeTypeSetting homeTypeSetting) {
        this.homeTypeSetting = homeTypeSetting;
    }

    @Override
    public void setTabMode(@MainContract.TabMode int mode) {
        MainViewState viewState;
        switch (mode) {
            case TAB_MODE_NOW:
                viewState = new MainViewState.NowState(homeTypeSetting.isVerticalType());
                break;
            case TAB_MODE_PLAN:
                viewState = new MainViewState.PlanState(homeTypeSetting.isVerticalType());
                break;
            case TAB_MODE_SETTINGS:
                viewState = MainViewState.SettingsState.INSTANCE;
                break;
            default:
                throw new IllegalStateException("Unknown tab mode");
        }
        view.render(viewState);
    }
}

File: app/src/main/java/soup/movie/ui/main/MainPresenter.kt
package soup.movie.ui.main

import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.internal.disposables.DisposableContainer
import soup.movie.data.MoopRepository
import soup.movie.settings.impl.LastMainTabSetting
import soup.movie.settings.impl.LastMainTabSetting.Tab
import soup.movie.settings.impl.LastMainTabSetting.Tab.*
import soup.movie.ui.BasePresenter
import soup.movie.ui.main.MainActionState.NotFoundAction
import soup.movie.ui.main.MainActionState.ShowDetailAction
import soup.movie.ui.main.MainViewState.*
import timber.log.Timber

class MainPresenter(private val lastMainTabSetting: LastMainTabSetting,
                    private val repository: MoopRepository) :
        BasePresenter<MainContract.View>(),
        MainContract.Presenter {

    override fun initObservable(disposable: DisposableContainer) {
        super.initObservable(disposable)
        disposable.add(lastMainTabSetting
                .asObservable()
                .distinctUntilChanged()
                .map { mapToViewState(it) }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { view?.render(it) })
    }

    private fun mapToViewState(tabMode: Tab): MainViewState =
            when (tabMode) {
                Now -> NowState
                Plan -> PlanState
                Theaters -> TheatersState
                Settings -> SettingsState
            }

    override fun setCurrentTab(mode: Tab) {
        Timber.d("setCurrentTab: %s", mode)
        lastMainTabSetting.set(mode)
    }

    override fun requestMovie(movieId: String) {
        //TODO: connect with view lifecycle
        repository.getMovie(movieId)
                .observeOn(AndroidSchedulers.mainThread())
                .map<MainActionState> { ShowDetailAction(it) }
                .defaultIfEmpty(NotFoundAction)
                .subscribe { view?.execute(it) }
    }
}
